Lib/test/test_io.py

```diff
@@ -3225,6 +3225,20 @@ def test_read_nonbytes(self):
         t = self.TextIOWrapper(self.StringIO('a'))
         self.assertRaises(TypeError, t.read)
 
+    def test_illegal_encoder(self):
+        # Issue 31271: Calling write() while the return value of encoder's
+        # encode() is invalid shouldn't cause an assertion failure.
+        class BadEncoder():
+            def encode(self, dummy):
+                return 42
+        def _get_bad_encoder(dummy):
+            return BadEncoder()
+        quopri = codecs.lookup("quopri")
+        with support.swap_attr(quopri, '_is_text_encoding', True), \
+             support.swap_attr(quopri, 'incrementalencoder', _get_bad_encoder):
```

Comment:
I am not sure I understood your question.
- if we remove it, then the assertion won't fail (even without the issue's patch).
- if we just overwrite quopri.incrementalencoder, without restoring it, the assertion would fail (without the issue's patch), but I am not sure about the effects of not restoring it (on Python with the issue's patch), when Python keeps running.
- can we reproduce the assertion failure without changing quopri.incrementalencoder? I don't know. I haven't looked for any other way, but I can do that, if you think it's worth it.

Reply:
What if use an existing encoding that produces non-bytes? E.g. "rot13"?

Reply:
indeed. I would update the patch.